zqd/ingest/log.go

```diff
@@ -5,207 +5,138 @@ import (
 // x509_20191101_14:00:00-15:00:00+0000.log.gz (corelight)
 const DefaultJSONPathRegexp = `([a-zA-Z0-9_]+)(?:\.|_\d{8}_)\d\d:\d\d:\d\d\-\d\d:\d\d:\d\d(?:[+\-]\d{4})?\.log(?:$|\.gz)`
 
-type readCounter struct {
-	f     *os.File
-	nread int64
-}
+type LogOp struct {
+	bytesRead  int64
+	bytesTotal int64
+	warnings   []string
+	readers    []zbuf.Reader
+	files      []*os.File
+	err        error
 
-func (rc *readCounter) Read(p []byte) (int, error) {
-	n, err := rc.f.Read(p)
-	rc.nread += int64(n)
-	return n, err
+	warningCh chan string
+	doneCh    chan struct{}
 }
 
-func (rc *readCounter) Close() error {
-	return rc.f.Close()
-}
-
-func openIncomingLog(path string) (*readCounter, int64, error) {
-	info, err := os.Stat(path)
-	if err != nil {
-		return nil, 0, err
-	}
-	if info.IsDir() {
-		return nil, 0, zqe.E(zqe.Invalid, "path is a directory")
-	}
-	f, err := fs.Open(path)
-	if err != nil {
-		return nil, 0, err
+// Logs ingests the provided list of files into the provided space.
+// Like ingest.Pcap, this overwrites any existing data in the space.
+func NewLogOp(ctx context.Context, store *storage.ZngStorage, req api.LogPostRequest) (*LogOp, error) {
+	p := &LogOp{
+		warningCh: make(chan string, 5),
+		doneCh:    make(chan struct{}),
 	}
-	return &readCounter{f: f}, info.Size(), nil
-}
-
-func ingestLogs(ctx context.Context, pipe *api.JSONPipe, s *space.Space, req api.LogPostRequest) error {
-	zctx := resolver.NewContext()
-	var readers []zbuf.Reader
-	defer func() {
-		for _, r := range readers {
-			if closer, ok := r.(io.Closer); ok {
-				closer.Close()
-			}
-		}
-	}()
-	cfg := detector.OpenConfig{}
+	var cfg detector.OpenConfig
 	if req.JSONTypeConfig != nil {
 		cfg.JSONTypeConfig = req.JSONTypeConfig
 		cfg.JSONPathRegex = DefaultJSONPathRegexp
 	}
-	var totalSize int64
-	var readCounters []*readCounter
+	zctx := resolver.NewContext()
 	for _, path := range req.Paths {
-		rc, size, err := openIncomingLog(path)
-		if err != nil {
-			return err
-		}
-		sf, err := detector.OpenFromNamedReadCloser(zctx, rc, path, cfg)
-		if err != nil {
-			if req.StopErr {
-				return fmt.Errorf("%s: %w", path, err)
-			}
-			pipe.Send(&api.LogPostWarning{
-				Type:    "LogPostWarning",
-				Warning: fmt.Sprintf("%s: %s", path, err),
-			})
-			continue
+		if err := p.openFile(zctx, path, cfg); err != nil {
+			return nil, err
 		}
-
-		totalSize += size
-		readCounters = append(readCounters, rc)
-		readers = append(readers, sf)
 	}
+	go p.start(ctx, store)
+	return p, nil
+}
 
-	logBytesRead := func() int64 {
-		var n int64
-		for i := range readCounters {
-			n += readCounters[i].nread
-		}
-		return n
-	}
+func (p *LogOp) openWarning(path string, err error) {
+	p.warnings = append(p.warnings, fmt.Sprintf("%s: %s", path, err))
+}
 
-	zngfile, err := s.CreateFile(filepath.Join(tmpIngestDir, allZngTmpFile))
+func (p *LogOp) openFile(zctx *resolver.Context, path string, cfg detector.OpenConfig) error {
```

Comment:
This function does too much, and all of the reader/writer plumbing that's going on is hard to follow. I think the former setup was easier to follow: loop over req.path, and do 3 separate things: open the file, calling the detector package to get a zbuf.Reader from it, then add the reader & info to the readers state.

Reply:
I agree that this function could be broken up. OTH IMHO the way it was was harder to follow. Instead you had the `ingestLogs` function that was doing too much.

Reply:
Reverted.